Data_Engine/Create/NTree.cs

```diff
@@ -0,0 +1,174 @@
+/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2020, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+using BH.oM.Base;
+using BH.oM.Geometry;
+using BH.oM.Reflection;
+using BH.oM.Data.Collections;
+
+namespace BH.Engine.Data
+{
+    public static partial class Create
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+        
+        public static NTree<T> NTree<T>(IEnumerable<NBound> boundingBoxes, IEnumerable<T> data, int splits = 16)
+        {
+            if (boundingBoxes.Count() != data.Count())
+                return null;
+
+            List<Leaf<T>> leaves = boundingBoxes.Zip(data, (x, y) => new Leaf<T>() { Bounds = x, Item = y }).ToList();
+
+            return NTree(leaves, splits);
+        }
+
+        /***************************************************/
+
+        public static NTree<T> NTree<T>(IEnumerable<Leaf<T>> leaves, int splits = 16)
```

Comment:
Looks like you are querying leaves as `IEnumerable` for its count a few times + there is a cast to `List` as well as to `Array`. Possibly could be done smarter?